### PR TITLE
Secure hosted chat runtime target propagation

### PR DESCRIPTION
**Pull request by kojiwakayama**
### Motivation
- Prevent client-controlled environment targeting by discarding `runtimeTargetKind` and `runtimeTargetEnvironmentId` from public hosted chat requests unless the request comes from a verified runtime invocation.
- Preserve existing functionality for trusted control-plane/runtime-signed invocations so runtime-target propagation still works when the caller is verified.

### Description
- Introduce an internal builder `buildParsedHostedChatRequestInternal` with a `trustRuntimeTargetMetadata` flag and export a public wrapper `buildParsedHostedChatRequest` that sets `trustRuntimeTargetMetadata: false`, ensuring public requests cannot inject environment targets (`src/agent/hosted/chat-request-parser.ts`).
- When parsing signed runtime invocations, call the internal builder with `trustRuntimeTargetMetadata: true` and keep the runtime source binding check via `verifyHostedRuntimeSourceBinding` to allow trusted propagation (`src/agent/hosted/chat-request-parser.ts`).
- Strip `runtimeTargetKind`/`runtimeTargetEnvironmentId` from untrusted chat context by extracting `publicChatContext` and using it for `validatedContext` when `trustRuntimeTargetMetadata` is false (`src/agent/hosted/chat-request-parser.ts`).
- Add regression tests covering both boundaries: public hosted chat requests must not carry environment targets, while verified runtime invocations retain them (`src/agent/hosted/chat-request.test.ts`).

### Testing
- Ran `git diff --check` which passed with no whitespace errors.
- Committed changes and updated tests; local `git status` shows the intended changes staged and committed.
- Attempted to run `deno` tests (`deno test --no-check --allow-all ...`) but `deno` is not available in the environment, and installation attempts were blocked (HTTP 403), so the new tests were added but could not be executed here.
- Manual verification performed via repository inspection and targeted edits to ensure the public parsing path does not propagate runtime target metadata while the runtime-invocation path does.

------
[Codex Task](https://chatgpt.com/codex/cloud/tasks/task_e_6a6fca4c80f4832d8529a5f46408dde7)

### CODE REVIEW

File: src/agent/hosted/chat-request-parser.ts [hunk 1]
```diff
@@ -493,13 +512,14 @@ export async function parseRuntimeAgentRunInvocationHostedChatRequestFromRequest
...
 
-  const parsedRequest = await buildParsedHostedChatRequest({
+  const parsedRequest = await buildParsedHostedChatRequestInternal({
     authToken: authenticatedRequest.authToken,
     userId: invocation.data.run.requestedByUserId,
     chatRequest: chatRequest.data,
     agentId: invocation.data.run.agentId,
     agentConfig: invocation.data.agentConfig,
     verifyProjectAccess: options.verifyProjectAccess,
+    trustRuntimeTargetMetadata: true,
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/hosted/chat-request-parser.ts]: Fixed in aae5913ba6. Runtime target metadata is now stripped during parsing and restored only after the run-event service token is cryptographically verified and bound to the parsed project/run; an ordinary authenticated invocation without that control-plane token stays on the public sanitized path. Added both verified and missing-token regressions. Pinned Deno 2.7.7 check passed and hosted-chat suite passed 72 steps.

**Comment by kwakayama** [comment]
## Codex security re-review — approve

Reviewed exact SHA `aae5913ba61f60aca47ef82f4cc866df37d6d48b` against base `d94a1cd9eb2d79bad969fb2c5a876ec858c15be1`.

No actionable findings.

Public hosted chat parsing always strips runtime target metadata. The runtime-invocation path restores it only after source binding and cryptographic verification of the run-event token bound to the parsed project/run; a missing token remains on the sanitized public path. The verified/missing-token regression pair exercises the security boundary directly.

### Verification

- Pinned Deno 2.7.7 `deno check --no-lock` over both changed TypeScript files: pass
- `src/agent/hosted/chat-request.test.ts`: 1 suite / 72 steps passed
- Changed-file formatting and `git diff --check BASE...HEAD`: pass
- Review threads: 1 total, 0 unresolved; pagination complete
- Exact-head check retrieval was valid and stable; no failing checks at review time, with external checks still queued
- `needs-human-input` remains present. This review does not authorize merge.

Score: 97/100

Review-Gate:
Reviewer: Codex
Reviewed-SHA: aae5913ba61f60aca47ef82f4cc866df37d6d48b
Score: 97/100
Actionable-Findings: 0
Verdict: APPROVE

File: src/agent/hosted/chat-request-parser.ts [hunk 2]
```diff
@@ -601,10 +617,23 @@ export async function parseRuntimeAgentRunInvocationHostedChatRequestFromRequest
...
     options.verifyRunEventAppendToken,
     true,
   );
+  if (verifiedRequest instanceof Response || !verifiedRequest.serverEnvelopeVerified) {
+    return verifiedRequest;
+  }
+
+  const runtimeContext = chatRequest.data.context;
+  return {
+    ...verifiedRequest,
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/hosted/chat-request-parser.ts]: Fixed in 3ff5ba0bb7ba19edfecb90e00ae7a9f0db9f93b7. The parser no longer returns a spread copy: it grafts the runtime target metadata onto the same `verifiedRequest` object the run-event token was registered against (mutating `validatedContext` in place), so the module-private WeakMap lookup in `createHostedRunEventWriterCapabilityForRequest` recovers the credential from the parser's actual return value. Added a regression test that asserts the capability is recoverable from the returned object (and that a spread copy correctly yields no capability, proving the credential stays bound to object identity). Verified the new assertion fails against the pre-fix code.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by kojiwakayama** [comment]
@codex review

**Comment by kwakayama** [comment]
@codex review

File: src/agent/hosted/chat-request-parser.ts [hunk 3]
```diff
@@ -601,10 +617,25 @@ export async function parseRuntimeAgentRunInvocationHostedChatRequestFromRequest
...
+  // keyed by the exact object identity of `verifiedRequest`, so the runtime
+  // target metadata is grafted onto that same object rather than a spread
+  // copy — returning a copy would orphan the credential lookup downstream.
+  const runtimeContext = chatRequest.data.context;
+  verifiedRequest.validatedContext = {
+    ...verifiedRequest.validatedContext,
+    runtimeTargetKind: runtimeContext.runtimeTargetKind,
+    runtimeTargetEnvironmentId: runtimeContext.runtimeTargetEnvironmentId,
+  };
+  return verifiedRequest;
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/hosted/chat-request-parser.ts]: Addressed in 821c5ab4. The verified runtime context is now constructed before registration, so the final returned request object is registered exactly once and no post-registration mutation or credential re-keying is needed. The focused parser suite confirms public and unverified requests strip runtime targets, verified control-plane requests preserve them, the exact returned object recovers the writer capability, and spread copies do not.

**Comment by kwakayama** [comment]
@codex review

**Comment by kwakayama** [comment]
@codex review
